### ISSUE TITLE
New common getExampleAlerts() method

### ISSUE THREAD

**Issue by psiinon** [zaproxy/zaproxy#6119]
This is for all scan rules - active, passive, http, websocket, future ones :)
The method is proposed to be a 'defacto standard' for now: `List<Alert> getExampleAlerts()`

It will be accessed by the [generate_alert_pages.js](https://github.com/zaproxy/zap-admin/blob/master/scripts/generate_alert_pages.js) script using introspection. This script generates the https://www.zaproxy.org/docs/alerts/ pages.

At the moment the script can only cope with one alert per add-on, while many addons generate serveral.

Ease of mainenance is key - this ne method should call any suitable existing methods - if they dont exist the new ones could be created.

Release & beta rules with little to no info which should implement this new method asap:

- [ ] https://www.zaproxy.org/docs/alerts/10003/ Vulnerable JS Library
- [ ] https://www.zaproxy.org/docs/alerts/10020/ X-Frame-Options Header
- [ ] https://www.zaproxy.org/docs/alerts/10032/ Viewstate
- [ ] https://www.zaproxy.org/docs/alerts/10036/ HTTP Server Response Header

The plan is to only expose the information currently on the alert pages, so URLs can be https://www.example.com

Any new generic text created for these alerts should be i18n'd

**Comment by kingthorin** [comment]
Is this going to be added to core and subsequently implemented by scan rules?

**Comment by psiinon** [comment]
Yeah, latest plan is to change the core interfaces, but I'll aim to implement the method for the rules in the first comment and we can see how well it works. The script already uses introspection so it should be able to cope without the core changes.
We can add a new 'ScanRule' interface with just this method (?) to the core then the other interfaces can extend it.

**Comment by kingthorin** [comment]
Sounds good.

One thing to keep in mind some i18n messages require replacements/insertions (I dunno what the proper term is). So dummy values will have to be passed for those. (The whole `some.key=Some interesting message about {0}` type thing 😉 )

**Comment by psiinon** [comment]
Yeah - thats what I meant by the i18n part - in case we need to introduce new more generic strings.

**Comment by psiinon** [comment]
Example of what one of the rules which raises multiple alerts would look like.
Currently its mostly blank :/ https://www.zaproxy.org/docs/alerts/10020/
![Screenshot_2020-08-13 OWASP ZAP(1)](https://user-images.githubusercontent.com/1081115/90150128-9cd7d200-dd85-11ea-9f36-802dfcb72faa.png)

**Comment by kingthorin** [comment]
Note: zaproxy/zaproxy#7100 may as well be tackled at the same time.

**Comment by kingthorin** [comment]
Updated, seem PII was the only one missed.

**Comment by kingthorin** [comment]
On all those that remain to be done at this point of added initials to clarify things a bit.
A - ascanrules
AB - ascanrulesBeta
AA - ascanrulesAlpha
P - pscanrules
PB - pscanrulesBeta
PA - pscanrulesAlpha

**Comment by iagoscm** [comment]
Hello! I'm looking upon to work on this issue with a friend of mine, we are trying to get into some repository on OWASP and this one is one of the most engaged, but we don't understand a lot of the field, and we confess we are a little bit lost on this, how could we start to work on it?

**Comment by kingthorin** [comment]
I'd suggest having a look at some of the PRs that have already contributed to this (see links/refs above). Then look at the rules that have work outstanding. Let us know which few you think you can tackle and we'll block them off for you. (I'd definitely suggest tackling singles or a small batch for your first contribution.)

**Comment by iagoscm** [comment]
Alright, I'm going to study and see what I can do. Thanks!

**Comment by kingthorin** [comment]
Please be sure to let us know if you decide to tackle some, so that we can ensure to mark them and not end-up with people tackling overlapping work :wink:

**Comment by iagoscm** [comment]
Hey @kingthorin, sorry for keep you waiting. I decided I'm going to try to tackle "PersistentXssScanRule"

**Comment by kingthorin** [comment]
Assigned.

You'll need to adjust:
- The [scan rule](https://github.com/zaproxy/zap-extensions/blob/main/addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/PersistentXssScanRule.java)
- The [unit tests](https://github.com/zaproxy/zap-extensions/blob/main/addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/PersistentXssScanRuleUnitTest.java)
- The [CHANGELOG](https://github.com/zaproxy/zap-extensions/blob/main/addOns/ascanrules/CHANGELOG.md)

**Comment by LucasBergholz** [comment]
Hello guys, would love to tackle this issue. Would you guys say any of the SQL Injection alerts are good for a first contribution on this issue?

**Comment by kingthorin** [comment]
Thought you were tackling "PersistentXssScanRule".

I haven't reviewed the SQLi rules recently.

**Comment by LucasBergholz** [comment]
@kingthorin I think you are confusing me with @iagoscm . Im new to this issue.

**Comment by kingthorin** [comment]
Oops sorry, you’re right.

**Comment by LucasBergholz** [comment]
@kingthorin I'll try to tackle SpringActuatorScanRule, could you assign it to me?

**Comment by kingthorin** [comment]
Added you.

**Comment by anaaroch** [comment]
Hello, I would like to try working with the pscanrule InsecureJsfViewStatePassiveScanRule, could you assign me this alert?

**Comment by kingthorin** [comment]
I'll set it Monday or Tuesday.

**Comment by Brandosp** [comment]
Hello! I was wondering if this is still something that could be worked on? I am not too familiar with the field but would love to learn and tackle one of the issues here if possible!

**Comment by kingthorin** [comment]
Anything with an ❌ should still be available.

To get a handle on what's needed just check any of the linked PRs in the issue's history.

**Comment by Brandosp** [comment]
Hello! I did a bit of research by looking at the linked PRs and I am interested in working on the InsecureHttpMethodScanRule if possible!

**Comment by thc202** [comment]
Set to you.

**Comment by sharvandeep** [comment]
Hi, I would like to work on this issue.  
I will add getExampleAlerts() to the relevant scan rules and open a PR soon.  

Thanks!  
@sharvandeep

**Comment by chenjingen-jane** [comment]
Is this still relevant? I'd like to help implement this for some common add-ons.

**Comment by kingthorin** [comment]
@chenjingen-jane yes. The embedded list is up-to-date. Just let us know which rule(s) you want to tackle and we can put your name on them.

**Comment by chenjingen-jane** [comment]
> [@chenjingen-jane](https://github.com/chenjingen-jane) yes. The embedded list is up-to-date. Just let us know which rule(s) you want to tackle and we can put your name on them.

Hi @kingthorin, thanks for the reply! I'd like to tackle the Session Fixation (40013) rule. Please put my name on it. I will start working on implementing the getExampleAlerts() method for it. Thanks!

**Comment by kingthorin** [comment]
Will do.

@chenjingen-jane I'd suggest having a look at some of the other PRs associated with this issue so you can see what needs to be done along with the expected Unit Test and CHANGELOG changes.

**Comment by Karl-Seryani** [comment]
Hi @kingthorin, I'd like to tackle **SQL Injection - SQLite (Time Based) (40024)**. I've opened a PR: https://github.com/zaproxy/zap-extensions/pull/7224

**Comment by Karl-Seryani** [comment]
I'd like to tackle UsernameEnumerationScanRule (40023) next. Will open a PR shortly.

**Comment by Karl-Seryani** [comment]
LDAP Injection (40015) - PR zaproxy/zap-extensions#7288

**Comment by sainadh777** [comment]
I’d like to work on adding example alerts for MongoDbInjectionScanRule (40033), including unit coverage and the add-on changelog.

**Comment by kingthorin** [comment]
Go for it!

**Comment by modithakolli** [comment]
Hi @kingthorin, I'd like to tackle SQL Injection - A (40018, SqlInjectionScanRule).  I'll follow the pattern from the other SQL injection rules (#7279) for reference. 
Thanks!

**Comment by kingthorin** [comment]
@modithakolli Thanks, go for it!
